state-chain/pallets/cf-elections/src/electoral_systems/solana_swap_accounts_tracking.rs

```diff
@@ -0,0 +1,214 @@
+					unsynchronised_state.witnessed_open_accounts.extend(
+						consensus.new_accounts.iter().map(|(account, swap_details)| {
+							Hook::initiate_vault_swap((*swap_details).clone());
+							(*account).clone()
+						}),
+					);
+
+					consensus.confirm_closed_accounts.into_iter().for_each(|acc| {
+						unsynchronised_state.closure_initiated_accounts.remove(&acc);
+					});
+
+					Ok(())
+				})?;
+			}
+		} else {
+			electoral_access.new_election((), (), ())?;
+		}
+
+		let mut unsynchronised_state = electoral_access.unsynchronised_state()?;
+		if Hook::get_number_of_available_sol_nonce_accounts() >
+			NONCE_AVAILABILITY_THRESHOLD_FOR_INITIATING_SWAP_ACCOUNT_CLOSURES &&
+			(unsynchronised_state.witnessed_open_accounts.len() >=
+				MAX_BATCH_SIZE_OF_CONTRACT_SWAP_ACCOUNT_CLOSURES ||
+				(*current_block_number)
+					.checked_sub(&unsynchronised_state.block_number_last_closed_accounts)
+					.expect(
+						"current block number is always greater than when apicall was last created",
+					)
+					.into() >= MAX_WAIT_BLOCKS_FOR_SWAP_ACCOUNT_CLOSURE_APICALLS)
+		{
+			let accounts_to_close: Vec<_> = if unsynchronised_state.witnessed_open_accounts.len() >
+				MAX_BATCH_SIZE_OF_CONTRACT_SWAP_ACCOUNT_CLOSURES
+			{
+				unsynchronised_state
+					.witnessed_open_accounts
+					.drain(..MAX_BATCH_SIZE_OF_CONTRACT_SWAP_ACCOUNT_CLOSURES)
+					.collect()
+			} else {
+				sp_std::mem::take(&mut unsynchronised_state.witnessed_open_accounts)
+			};
+			match Hook::close_accounts(accounts_to_close.clone()) {
+				Ok(()) => {
+					unsynchronised_state.block_number_last_closed_accounts = *current_block_number;
+					unsynchronised_state.closure_initiated_accounts.extend(accounts_to_close);
+					electoral_access.set_unsynchronised_state(unsynchronised_state)?;
+				},
+				Err(e) => {
+					log::error!(
+						"failed to build Solana CloseSolanaVaultSwapAccounts apicall: {:?}",
+						e
+					);
+				},
+			}
+		}
+		Ok(())
+	}
+
+	fn check_consensus<ElectionAccess: ElectionReadAccess<ElectoralSystem = Self>>(
+		_election_identifier: ElectionIdentifier<Self::ElectionIdentifierExtra>,
+		_election_access: &ElectionAccess,
+		_previous_consensus: Option<&Self::Consensus>,
+		consensus_votes: ConsensusVotes<Self>,
+	) -> Result<Option<Self::Consensus>, CorruptStorageError> {
+		let num_authorities = consensus_votes.num_authorities();
+		let success_threshold = success_threshold_from_share_count(num_authorities);
+		let active_votes = consensus_votes.active_votes();
+		let num_active_votes = active_votes.len() as u32;
+		Ok(if num_active_votes >= success_threshold {
+			let mut counts = BTreeMap::new();
+			for vote in active_votes {
+				counts.entry(vote).and_modify(|count| *count += 1).or_insert(1);
+			}
+			counts.iter().find_map(|(vote, count)| {
+				if *count >= success_threshold {
+					Some(vote.clone())
+				} else {
+					None
+				}
+			})
+		} else {
+			None
+		})
+	}
```

Comment:
this is the same in 3 electoral systems, should look at factoring this out

Reply:
the impl has changed now and so it is not the same as other electoral systems.